gateway/ln-gateway/src/ng/register.rs

```diff
@@ -0,0 +1,149 @@
+use std::time::Duration;
+
+use fedimint_client::sm::{OperationId, State, StateTransition};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::task::sleep;
+use fedimint_ln_common::api::LnFederationApi;
+use fedimint_ln_common::LightningGateway;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::error;
+use url::Url;
+
+use super::GatewayClientContext;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// State machine that registers the lightning gateway with
+/// the federation.
+///
+/// ```mermaid
+/// graph LR
+/// classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///    Register -- register lightning gateway succeeded --> WaitForTTL
+///    Register -- register lightning gateway failed --> Register
+///    WaitForTTL -- wait for time to live to expire --> Register
```

Comment:
Re-registration after TTL seem like a good opportunity to update the gateway info registered with the Fed. for example gateway might want to update it's node pubkey.

- How would an existing SM know of this info update so it can forward it on next registration?
- Might an existing Register SM be canceled and replaced by a newer one, even before TTL expires?

Reply:
The new implementation of `transition_wait_for_ttl` should make these situations easy. See this comment https://github.com/fedimint/fedimint/pull/2628#discussion_r1224746331

The high level idea is that if the `LightningGateway` registration info is updated in the local db, this state machine will stop, but that means a new one was created which will take care of the registration. The old one just waits for the TTL to expire and then it exits. 

Un-registering a gateway hasn't been implemented yet but it should be trivial. Just delete the `FederationRegistrationKey` from the database.

Reply:
> Un-registering a gateway hasn't been implemented yet but it should be trivial. Just delete the FederationRegistrationKey from the database.

In which case, a registered gateway would continue to serve until TTL expires right?